tiledb/sm/group/group_directory.h

```diff
@@ -134,10 +134,10 @@ class GroupDirectory {
   URI uri_;
 
   /** The storage manager. */
-  VFS* vfs_;
+  VFS& vfs_;
 
   /** A thread pool used for parallelism. */
-  ThreadPool* tp_;
+  ThreadPool& tp_;
```

Comment:
I must say I have felt a bit uneasy by using raw references and pointers in places like this, because of potential lifetime management concerns (which might cause problems in the future). What would the performance impact be if these were shared pointers? They don't have any overhead when just dereferenced, do they?

Reply:
There is an overhead. But in the case of VFS, if you're going to do IO after, I doubt this is going to be significant. For the thread pool, it's also heavy so I wouldn't be concerned about a shared pointer dereference.

Reply:
Note that changing from references to pointers doesn't introduce any more lifetime than there is already. The reference can also be deleted before the use here are fully done. So this PR doesn't introduce anything and probably can proceed, but it would be good to have a better story around contexts going away before an operation is done. Let's file a story here as this will be more involved.

Reply:
@eric-hughes-tiledb thoughts?

Reply:
Doesn't this call shared_ptr->get()?

Reply:
What do you mean by "this"?

Reply:
Actually, ->get() will call the dereference operator so it's the other way around... We have had instances where calling ->get() will cause performance impact so the dereference operator will cause the same. The getter for a shared pointer has to be protected at some point for multi threading and that protection is not free.

Reply:
The [dereference operators call `get()`](https://github.com/microsoft/STL/blob/ff0cff1ad6de63525d0a6646b49cc10667446682/stl/inc/memory#L1761-L1774) which simply reads a pointer field as shown above. The reference count is only altered when a shared pointer gets created or destroyed.

Reply:
"simply reads a pointer field": I think the pointer field ends up being an atomic which is guarded by a mutex. I didn't have time to read the full code, but we had a HUGE performance regression when switching something to a shared_ptr and that code was calling the arrow operator on a shared_ptr which also "simply reads a pointer field", so my gut tells me this is no different here. It's not about the ref counting but some protection on the ptr.

Reply:
[The pointer field is not an atomic.](https://github.com/microsoft/STL/blob/ff0cff1ad6de63525d0a6646b49cc10667446682/stl/inc/memory#L1447) I remember you having mentioned these performance regressions, but IIRC these were by using shared pointers in parameters or in short-lived objects. `GroupDirectory` is relatively long-lived so it would be less susceptible to similar issues, but it would be good to make some performance measurements before making any changes (I'm not saying we should use shared pointers right now).

Reply:
No, it was by accessing functions on dimensions, which were stored in a shared pointer, not by sending shared pointers down. Make sure you follow the inheritance properly. There has got to be some way that a shared pointer can be updated safely in a multi threaded environment, and I do see some atomic inheritance in that file.

Reply:
@eric-hughes-tiledb do you recall why we were seeing a performance issue when we switched over to shared_ptr for dimensions? I might be wrong about multi threading protection but I recall there was something there that made us cache the raw pointers?

Reply:
>something there that made us cache the raw pointers

It was performance. We cannot rely on the particular behavior of MSVC or any other single compiler to make this kind of decision.

>potential lifetime management concerns 

We have made a hard requirement that almost everything in the code operates under an instance of `class Context`, which means that the lifespan of these resources is guaranteed to be longer than the operations. What's not happening under `Context` are the handful of things that set up `Context` and these don't require such resources.

All this is inadequately documented IMO. Once we're done gutting `class StorageManager` we'll be in a good place to make such documentation and be able to audit compliance with it.

We may have some minor issues with task cancellation etc., but that system needs to be rewritten regardless.

Reply:
I'm resolving this conversation because it's the wrong venue for the concern raised. The lifespan issues are the same for both references and pointers (as well as `std::reference_wrapper`) so the issue is just not relevant to the current PR. The same issue is present in both the "before" and "after" versions, so the concern isn't going to change anything about this PR. The scope is too large to consider making any such change here.

 Simply because some code is being changed doesn't mean that it's a good place to talk about something that's possibly relevant.